ykcompile/src/lib.rs

```diff
@@ -208,23 +209,87 @@ impl<TT> TraceCompiler<TT> {
+                            Location::Mem(ro) => {
+                                // Deref value and copy it.
+                                dynasm!(self.asm
+                                    ; mov Rq(temp), [Rq(ro.reg) + ro.offs]
+                                    ; mov Rq(temp), [Rq(temp)]
+                                );
+                            }
+                            Location::Register(reg) => {
+                                dynasm!(self.asm
+                                    ; mov Rq(temp), [Rq(reg)]
+                                );
+                            }
+                            _ => unreachable!(),
+                        };
+                        ty = SIR.ty(&tyid).clone();
+                        curloc = Location::Register(temp)
+                    } else {
+                        // FIXME If the `Deref` is followed by an `Index` or `Field` projection, we
+                        // just let them handle this reference. Otherwise, we need to copy the
+                        // array/struct/tuple to the stack.
+                    }
+                }
+                Projection::Index(local) => {
+                    // Get the type of the array elements.
+                    let elemsize = match ty {
+                        Ty::Array(_) => {
+                            // FIXME Since we can't compile array construction yet, we can assume
+                            // we are always dealing with references here. Once we can, the
+                            // assembler instructions below also need updating.
+                            todo!()
+                        }
+                        Ty::Ref(tyid) => match SIR.ty(&tyid) {
+                            Ty::Array(ety) => SIR.ty(&ety).size(),
+                            _ => unreachable!(),
+                        },
+                        _ => unreachable!(),
+                    };
+                    // Compute the offset of this index.
                     let temp = self.create_temporary();
-                    match curloc {
+                    match self.local_to_location(*local) {
+                        Location::Register(reg) => {
+                            dynasm!(self.asm
+                                ; imul Rq(temp), Rq(reg), elemsize as i32
+                            );
+                        }
+                        Location::Mem(ro) => {
+                            dynasm!(self.asm
+                                ; imul Rq(temp), [Rq(ro.reg) + ro.offs], elemsize as i32
+                            );
+                        }
+                        _ => todo!(),
+                    }
+                    // Add together the index and the array address and retrieve its value.
+                    match &curloc {
                         Location::Mem(ro) => {
                             dynasm!(self.asm
-                                ; mov Rq(temp), [Rq(ro.reg) + ro.offs]
```

Comment:
So looks like it wasn't practical to use `lea`?

Reply:
I'm more than happy if you have suggestions on how to do this stuff in less instructions. But I don't think `lea` is gonna help us here.